README.md
# asynccom-windows
This README file is best viewed [online](http://github.com/commtech/asynccom-windows/).

## Installing Driver

##### Downloading Driver Package
You can download a pre-built driver package directly from our [website](https://fastcomproducts.com/software/).


## Quick Start Guide
There is documentation for each specific function listed below, but lets get started with a quick programming example for fun. _This tutorial has already been set up for you at_ [`asynccom-windows/examples/tutorial/tutorial.c`](https://github.com/commtech/asynccom-windows/blob/master/examples/tutorial/tutorial.c).

Create a new C file (named tutorial.c) with the following code.

```c
#include <stdio.h>
#include <stdlib.h>
#include <Windows.h>
#include <ntddser.h>
#include "asynccom.h"

int main(void)
{
	HANDLE h = 0;
	DWORD tmp;
	unsigned char odata[] = "Hello world!";
	unsigned char idata[20];
    DCB mdcb;
    BOOL success;

	/* Open port 0 in a blocking IO mode */
	h = CreateFile(L"\\\\.\\COM8", GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
	if (h == INVALID_HANDLE_VALUE) {
		fprintf(stderr, "CreateFile failed with %d\n", GetLastError());
		return EXIT_FAILURE;
	}
	
    memset(&mdcb, 0, sizeof(mdcb));
    mdcb.DCBlength = sizeof(mdcb);
    success = GetCommState(h, &mdcb);
    if (!success)
    {
        printf("GetCommState failed! %d\n", GetLastError());
        return EXIT_FAILURE;
    }

    mdcb.BaudRate = 115200;
    mdcb.ByteSize = 7;
    mdcb.Parity = NOPARITY;
    mdcb.StopBits = ONESTOPBIT;
    if (SetCommState(h, &mdcb) == FALSE) {
        fprintf(stderr, "SetCommState failed with %d\n", GetLastError());
        return EXIT_FAILURE;
    }
    PurgeComm(h, PURGE_TXCLEAR | PURGE_RXCLEAR);


	WriteFile(h, odata, sizeof(odata), &tmp, NULL);

	/* Read the data back in (with our loopback connector) */
	ReadFile(h, idata, sizeof(idata), &tmp, NULL);

	fprintf(stdout, "%s\n", idata);

	return 0;
}
```

For this example I will use the Visual Studio command line compiler, but you can use your compiler of choice.

```
# cl tutorial.c
```

Now attach the included loopback connector.

```
# tutorial.exe
Hello world!
```

You have now transmitted and received data!


## Using The Serial Port
### Setting Baud Rate
##### Max Supported Speeds
The Fastcom: Async Com has a maximum baud rate 40 Mhz, but throughput may be lower than that.


The Fastcom: Async Com cards have their baud rate configured using the standard Windows API, such as IOCTL_SERIAL_SET_BAUD_RATE.

To get a non-standard baud rate there are a couple variables you need to setup before you can achieve them.

First is the variable clock generator frequency and second is the variable sampling rate. The formula for determining a baud rate is as follows.

```
Baud Rate = Clock Rate / Sampling Rate / Integer Divisor.
```

The 'Integer Divisor' value is determined in the driver and as long as the rest of the formula allows for an integer divisor it can be ignored.

Here is an example of some values that will work. We would like a baud rate of 1 MHz so we find a combination of a clock rate of 16 MHz and a sampling rate of 16 that can be divided by an integer to end up with 1 MHz. Now if we configure these two values we will be able to achieve any supported rate we want.

```
1,000,000 = 16,000,000 / 16 / 1
```

Once you understand how to achieve a desired baud rate, you can use our examples to achieve whatever baud rates you need.


## API Reference

There are likely other configuration options you will need to set up for your  own program. All of these options are described on their respective documentation page.

- [Connect](docs/connect.md)
- [Clock Rate](docs/clock-rate.md)
- [Echo Cancel](docs/echo-cancel.md)
- [External Transmit](docs/external-transmit.md)
- [Frame Length](docs/frame-length.md)
- [Isochronous](docs/isochronous.md)
- [9-Bit Protocol](docs/nine-bit.md)
- [Read](docs/read.md)
- [RS485](docs/rs485.md)
- [RX Trigger](docs/rx-trigger.md)
- [Sample Rate](docs/sample-rate.md)
- [TX Trigger](docs/tx-trigger.md)
- [Write](docs/write.md)
- [Disconnect](docs/disconnect.md)



### FAQ

## Build Dependencies
- Windows Driver Kit (7.1.0 used internally to support XP)


## Run-time Dependencies
- OS: Windows 7+


## API Compatibility
We follow [Semantic Versioning](http://semver.org/) when creating releases.


## License

Copyright (C) 2019 [Commtech, Inc.](https://fastcomproducts.com/)

Licensed under the [MIT license](https://opensource.org/licenses/MIT).


ChangeLog.md
# ChangeLog

## [2.0.10](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.10) (08/16/2019)
- Reworked the internal request handling to support other APIs and terminal programs.

## [2.0.9](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.9) (06/21/2019)
- Made the IOCTL queue parallel, allowing more terminal program support.
- Partially implemented GET_WAIT_MASK, SET_WAIT_MASK, and WAIT_ON_MASK.

## [2.0.8](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.8) (05/31/2019)
- Fixed a bug that caused reading and writing to be slower than intended.

## [2.0.7](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.7) (04/02/2019)
- Updated to the MIT license.

## [2.0.6](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.6) (02/25/2019)
- Fixed the 'reprogram firmware' IOCTL.
- Updated the copyright years.

## [2.0.5](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.5) (04/27/2018)
- Removed references to nine bit mode. it is not currently supported. References in the driver are retained for backwards compatibility purposes only.
- Changed IntervalTimeout to immediately occur, increasing compatibility with terminal programs.
- Change /inc to /lib/raw to be more in line with other Commtech repositories.

## [2.0.4](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.4) (09/01/2017)
- Finished implementing isochronous.
- Modified the naming to hopefully resolve some character size issues.

## [2.0.3](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.3) (07/03/2017)
- Modified the drivers to utilize 10.0 Driver Framework for Windows 10, and utilize non-executable memory.
- Some documentation tweaks.
- Fixed the build configurations.
- Small changes to the INF.

## [2.0.2](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.2) (04/19/2017)
- A special release for our new Vendor ID. All releases going forward use this new ID, 0x2EB0.

## [2.0.1](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.1) (02/08/2017)
- Fixed supporting multiple cards.

## [2.0.0](https://github.com/commtech/asynccom-windows/releases/tag/v2.0.0) (01/23/2017)
- Added COM compatibility to the drivers. Not backwards compatible with previous software.

## [1.0.1](https://github.com/commtech/asynccom-windows/releases/tag/v1.0.1) (01/17/2016)
- Fixed some errors in the .INX for Windows 10.

## [1.0.0](https://github.com/commtech/asynccom-windows/releases/tag/v1.0.0) (09/26/2016)
This is the initial release of the 1.0 driver series.

inc/asynccom.h
/*
Copyright (C) 2016  Commtech, Inc.

This file is part of asynccom-windows.

asynccom-windows is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

asynccom-windows is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with asynccom-windows.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef ASYNCCOM_PUBLIC_H
#define ASYNCCOM_PUBLIC_H

#ifdef __cplusplus
extern "C"
{
#endif
#include <string.h>

struct synccom_memory_cap {
	int input;
	int output;
};

#define ASYNCCOM_IOCTL_MAGIC 0x8020
#define ASYNCCOM_IOCTL_INDEX 0x800

#define IOCTL_ASYNCCOM_ENABLE_RS485 CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+0, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_DISABLE_RS485 CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+1, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_RS485 CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+2, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_ENABLE_ECHO_CANCEL CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+3, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_DISABLE_ECHO_CANCEL CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+4, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_ECHO_CANCEL CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+5, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_SET_SAMPLE_RATE CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+6, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_SAMPLE_RATE CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+7, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_SET_TX_TRIGGER CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+8, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_TX_TRIGGER CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+9, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_SET_RX_TRIGGER CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+10, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_RX_TRIGGER CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+11, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_SET_CLOCK_RATE CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+12, METHOD_BUFFERED, FILE_ANY_ACCESS)



#define IOCTL_ASYNCCOM_SET_EXTERNAL_TRANSMIT CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+16, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_EXTERNAL_TRANSMIT CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+17, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_SET_FRAME_LENGTH CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+18, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_FRAME_LENGTH CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+19, METHOD_BUFFERED, FILE_ANY_ACCESS)

#define IOCTL_ASYNCCOM_ENABLE_9BIT CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+20, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_DISABLE_9BIT CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+21, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_GET_9BIT CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+22, METHOD_BUFFERED, FILE_ANY_ACCESS)

// fixed baud rate

#define IOCTL_ASYNCCOM_REPROGRAM CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+26, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_ASYNCCOM_SET_DIVISOR CTL_CODE(ASYNCCOM_IOCTL_MAGIC, ASYNCCOM_IOCTL_INDEX+27, METHOD_BUFFERED, FILE_ANY_ACCESS)









#ifdef __cplusplus
}
#endif
#endif
